### DISCUSSION TITLE
Support for REC BMS via RS485?

### DISCUSSION THREAD

**Discussion by jonasf21** [arendst/Tasmota#15696] [Ideas]
Hello,

i am looking to find a way to read data from my REC BMS Q and send it via MQTT by Tasmota. The REC BMS Q has a RS232-style Interface on the back but uses RS485 protocol [1]. It appears to me that this is not ModBus as the packets are similar but seem not to conform to ModBus standards. I am not a ModBus pro, so maybe someone can clarify?
With Tasmota and a ESP8266 (or 32) and a TTL to RS485 converter its possible to send/receive data to/from the REC BMS. For example its possible to send the command "LCD1?" which will return most important values. The command has to be sent in hex (big endian) and requires a crc and a special ending. For this static command the full payload would be: 
55 01 00 05 4C 43 44 31 3F 46 D0 AA
The BMS will answer then - and this is where i am struggling. It will return data like this:

55 00 01 02 32 38 8E B4 AA 55 00 01 1C 67 66 7E 40 0E 62 7F 40 00 00 00 00 00 00 9C 41 07 41 6F 42 85 EB 51 3F B6 73 7A 3F      EE F0 AA

Its possible to decode this (see manual, page 13), the first part of the answer is useless then comes the important part which is all little endian hex float:
67 66 7E 40 min vcell volt (=little endian hex float = 3,975 Volt here)
0E 62 7F 40 max.cell volt
00 00 00 00 current A
00 00 9C 41 max temp.
07 41 6F 42 pack voltage
85 EB 51 3F state of charge
B6 73 7A 3F state of health
EE F0 checksum (CRC16)
AA (packets always end with this)

I see that tasmota scripting supports hex float to number string via hf("hexfloat" r) - this would convert the hex part to human readable data. 
Now the problem: how do i split this string to get an array or alike to have it exported via MQTT? Best would probably to write a library/driver to support all REC features (and include CRC16 for all commands etc) but i am nowhere near familar with that.

BR,
Jonas

[1] https://www.rec-bms.com/datasheet/UserManual_REC_SI_BMS.pdf

**Comment by barbudor** [comment]
Hi @jonasf21 

> With Tasmota and a ESP8266 (or 32) and a TTL to RS485 converter its possible to send/receive data

Yes but you need to use the recommended converter which automaticaly handle the direction of the converter as Tasmota doesn't have a provision to drive the direction pin. Supported converter include this module:
https://www.aliexpress.com/item/1005001636576393.html
Note I choose the link not for the price but for the good photo which helps recognize the supported module. Supported module only have VCC, GND, TX and RX on the ESP side with any direction control pin.

I'm not a specialist of Scripting-Language and Smart-Meter interface but with the syntax described [here](https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-metrics) you can apparently skip bytes with the `xN` token such as `x13` to ignore the first 13 bytes.
Token like `UUuuUUuu` or `uuUUuuUU` will extract 32 bytes unsigned long depending on endianess.
May be @gemu2015 could help with that.


Alternative to Scripting-Language is to use an ESP32 with Berry language which is close to Python in syntax and usage (although not 100% identical) which support Serial port so you could write a driver for that and publish to MQTT the data extracted.
Note as of today you can't write a full Tasmota-Energy compliant in Berry (no ENERGY reporting in SENSOR and accumulation per day and total), but you can do you own MQTT Publishing from the driver.

Hope this provide some clues

**Reply by pkkrusty** [reply]
@jonasf21 If you move forward with this, will be interested in hearing your progress.

**Comment by jonasf21** [comment]
@barbudor yes, i am using exactly this module.

Thanks for your suggestions - i'll see if i can get something running with that (i'll be happy with atleast basic info from the BMS).

Meanwhile i've found a project for powershell that does exactly what i want (export data received via serial to MQTT):
[https://github.com/aaronsb/REC-BMSManager](url)

Maybe someone could use this work to get it "ported" to tasmota.

**Reply by s-hadinger** [reply]
I don't know well enough Modbus but I feel that you can easily use serial interface provided by Berry, like @barbudor mentioned.

**Comment by gemu2015** [comment]
@jonasf21 
this is definitely not compatible with MODBUS as implemented in the SML driver.
not even the checksum is modbus compatible. this is a very inconsistent implementation.


i added a cmd to scripter to send serial cmds to this device with the appropriate checksum.

i don't know your programming skills but here is a skeleton to start with

\>D
res=0

; transmit array
m:trx_a=0 30
; receive array
m:rec_a=0 120

\>B
; preset cmd
trx_a[1]=0x55
trx_a[2]=0x01
trx_a[3]=0x00
trx_a[4]=0x05
trx_a[5]=0x4C
trx_a[6]=0x43
trx_a[7]=0x44
trx_a[8]=0x31
trx_a[9]=0x3F
trx_a[10]=0x00
trx_a[11]=0x00
trx_a[12]=0xAA

; open serial port, replace pin numbers with your actual wiring recpin, trxpin and baudrate
res=so(19 18 9600 8N1 1024)

\>S
; check if serial port received bytes
res=sa()
if res>0 {
; read result into array
res=sra(rec_a);
print %res% bytes received
; now analyze array

if rec_a[1]==0x55 {
print valid receive
; now analyze array for results
}

}


; send next cmd array with checksum
swa(trx_a 12 1)


#define USE_SCRIPT_SERIAL
#define USE_SML_M
#define USE_SCRIPT

cmd af(array) converts first 4 byte array cells to float

**Comment by gemu2015** [comment]
here is a more complete version to show how to implement such a device, it currently only decodes one cmd: LCD1

\>D

res=0

; transmit array
m:trx_a=0 30
; receive array
m:rec_a=0 120

minv=0
maxv=0
curr=0
mtmp=0
pvolt=0
soc=0

\>B
; preset cmd
trx_a[0]=0
trx_a=0x55
trx_a=0x01
trx_a=0x00
trx_a=0x05
trx_a=0x4C
trx_a=0x43
trx_a=0x44
trx_a=0x31
trx_a=0x3F
trx_a=0x00
trx_a=0x00
trx_a=0xAA

; open serial port, replace pin numbers with your actual wiring recpin, trxpin and baudrate
res=so(19 18 9600 8N1 1024)

\>S
; check if serial port received bytes
res=sa()
if res>0 {
; read result into array
res=sra(rec_a);
print %res% bytes received

if rec_a[1]==0x55 {
print valid receive
; now analyze array for results
minv=af(rec_a 4)
maxv=af(rec_a 8)
curr=af(rec_a 12)
mtmp=af(rec_a 14)
pvolt=af(rec_a 18)
soc=af(rec_a 22)
}

}

; send next cmd array with checksum
swa(trx_a 12 1)

; MQTT export
\>J
,"BMS":{
"minv":%2minv%,
"maxv":%2maxv%,
"curr":%2curr%,
"mtmp":%2mtmp%,
"pvolt":%2pvolt%,
"soc":%2soc%

; WEB display
\>W
Min Cell Voltage{m}%minv% V
Max Cell Voltage{m}%maxv% V
Min Current{m}%curr% A
Max Temp{m}%mtmp% C
Pack Voltage{m}%pvolt% V
SOC{m}%soc% %%


#define USE_SCRIPT
#define USE_SCRIPT_SERIAL
#define USE_SML_M
#define USE_SCRIPT_WEB_DISPLAY
#define USE_SCRIPT_JSON_EXPORT

**Reply by laurentppol** [reply]
Could You post example of how can I adapt Your script to my KL-F battery monitoring meter? It communicates over serial at 115200 8N1 (RS485) using something like this protocol (simple query string, returning all measured values), but response is series of comma separated decimals. Most of them need to be divided by 10 or by 100. For this speed I assume I have to use HW serial?
[KG-F User Manual.pdf](https://github.com/arendst/Tasmota/files/8784570/KG-F.User.Manual.pdf)

**Reply by pkkrusty** [reply]
This is great @gemu2015 ! Thank you for the starting point! One question, as I try to adapt this to various Modus devices, can the 8N1 be changed in `res=so(19 18 9600 8N1 1024)` or is it stuck there due to software serial instead of hardware? Or does that only apply to the Serial->TCP driver?

**Comment by gemu2015** [comment]
@laurentppol 

this is an example for the pylon tech battery monitor which also uses plain asci
it should be very easy to adapt this to your meter

\>D 64
res=0
rstr=""
label=""
lraw=""
val=0
volt=0
curr=0
temp=0
coul=0

\>B
; open serial port
res=so(3 1 115200 8N1 1024)

\>F
; check for serial data
res=sa()
if res>0 {
; read until linefeed
rstr=sr(10)
; decompose string
rstr=sb(rstr 1 sl(rstr))
label=sb(rstr 0 7)
lraw=sb(rstr 0 10)

val=st(rstr : 2)

switch label
case "Voltage"
volt=val/1000
case "Current"
if lraw=="Current   " {
curr=val
}
case "Tempera"
temp=val/1000
case "Coulomb"
coul=val
ends

}


\>S
; send every 5 seconds
if upsecs%5==0 {
res=sw("pwr 1\n")
}

\>W
Spannung{m}%4volt% V
Strom{m}%0curr% mA
Temperatur{m}%3temp% C
Ladung{m}%0coul% %%

\>J
,"pylon":{"Voltage":%4volt%,"Current":%0curr%,"Coulomb":%0coul%,"Temperature":%3temp%}

@pkkrusty 
8N1 can be changed as needed.  however you should use hardware serial on other formats then 8N1
for standard MODBUS devices i would use the SML driver. there are many examples of various drivers in the docs.
these are controlled by descriptors not by program

**Reply by pkkrusty** [reply]
@gemu2015 You mentioned that you added a command to scripter, and attached the xdrv_10_scripter.ino.zip file.  I assume all this is included in PR #15699 that you submitted yesterday?

**Reply by gemu2015** [reply]
Yes, but I added one extension for the extraction of rec bms , indexed array extract
So if you do not need rec bms you may use my latest pr
I will do a new pr if someone with the real hardware has tested

**Reply by pkkrusty** [reply]
I need to dive into the LTC6802-1 BMS chip, that communicates over SPI, and see that SPI support is part of scripter `USE_SCRIPT_SPI`. Is there anywhere documented the commands used with SPI?

However, the guy I work with deals with RecBMS, so I may play with the shell you put together above. I'll wait for that PR if you are planning on it.

**Reply by gemu2015** [reply]
@pkkrusty 
SPI is supported by scripter, either hardware or software spi. the LTC6802 needs to serve a lot of registers, should be doable however.
scripter docs show the spi cmds supported.

regarding MODBUS we now have full support for standard and REC-BMS transmitting and receiving.
Also MODBUS client mode is supported

**Reply by gemu2015** [reply]
@pkkrusty 
just made a pr with newest additions
here is a small example of an SPI usage

\>D
res=0
cnt=0
m:cmds=0 30

\>B
; define software SPI
; sclk = 5, mosi = 21; miso = 19
res=spi(0 5 21 19)
; device 1 cs = 15
res=spi(1 1 15)

; define array to transfer, read 6802 config registers
cmds[0]=0
cmds=0x02
cmds=0xff
cmds=0xff
cmds=0xff
cmds=0xff
cmds=0xff
cmds=0xff
: transfer 7 bytes to SPI device
res=spi(2 1 cmds 7 1)

; now in cmds may evaluate received data
for cnt 2 7 1
print %hn(cmds[cnt])%
next

**Reply by pkkrusty** [reply]
Amazing, thank you!

**Reply by pkkrusty** [reply]
A few clarification questions:

In your example, four bytes are stored in the array, correct?

Then in the transfer `res=spi(2 1 cmds 4 1)` , all four commands are transferred because of the 4? I'm confused how the array and the 4 interact.

In my case, the LTC6802-1 will send back an array of 6 bytes if I send 0x02 command, would I need to limit my send to 1 byte in that case?

So I do the following:

```
; define array to transfer
cmds[0]=0
cmds=0x02

; transfer 1 byte to SPI device
res=spi(2 1 cmds 1 1)
```
and res will have the returned array?

**Reply by gemu2015** [reply]
SPI always transmits and receives concurrently.
so if you want to receive 6 bytes you will have to send 6 bytes. i assume your command is only 0x02 with no parameters.
so you will have to send 7 bytes in total. normally the byte values send on the receive part are don't care but mostly we send a filler byte 0xff in that case.
so you would send 0x02 and 6 0xff bytes and then examine the array result from byte 2 to byte 7

**Reply by gemu2015** [reply]
res returns number of bytes sent.  in spi mode the array is IO that means the cmds input array contains the result after sending
i modified the above example to display the config registers of your device

**Reply by pkkrusty** [reply]
A few more questions regarding array syntax:

In 
```
cmds[0]=0
cmds=0x02
cmds=0xff
cmds=0xff
cmds=0xff
cmds=0xff
cmds=0xff
cmds=0xff
```
cmds[0]=0 clears the entire array?

normally first entry in the array is identified by array[1]?

In your example you write multiple 8-bit values to cmds array without specifying place ([1], [2], etc). In what cases is this allowed?

**Reply by gemu2015** [reply]
Array indices go from 1 .. N
Array[0] contains an array pointer
Array[-1] returns array length
If you store a value into an array without Index the value is stored at index and the index is incremented (auto increment mode)
If index reaches end of array it is reset to zero, wraps around

Arrays originally were filters, either moving average or median, I then decided to expand functionality to arrays
So storing a value into a filter array calculates a moving average
Reading the filter gives the moving average 

If you are in doubt simply use index mode only array[x]

**Reply by pkkrusty** [reply]
One more question: math operations are allowed in scripting according to docs: `Math operators +,-,*,/,%,&,|,^,<<,>>`

The carat seems to be included in the bitwise operators section as XOR, correct? For exponents, we should use pow(x y)? 

If I do the following:
```
numcells=12
mci=pow(2 numcells)-1
temp1=pow(2 2)
print DCC = %dcc%, MCI = %mci%, Min = %min%, Max = %max%, temp1 = %temp1%, numcells = %numcells%
```
I get:
```
19:32:32.448 DCC = 0.00, MCI = 3976.25, Min = 3.69, Max = 3.75, temp1 = 3.88, numcells = 12.00
```
I would expect pow(2 2) to give 4 and pow(2 numcells)-1 to give 4095

**Reply by gemu2015** [reply]
the exact power function uses a lot of flash memory and computing time. so it was decided by the team to replace it by an approximation which gives only results in the near of the real value. i will add a remark in the docs.

**Reply by pkkrusty** [reply]
I guess it's pretty easy to just make an array with the binary exponents in it, and just refer to it by index.

**Reply by gemu2015** [reply]
If you only need powers of 2 the easiest way is a simple shift
1<<n

**Reply by pkkrusty** [reply]
Yep, that's even better. 

Now I am running into a problem using the mean/median. I have defined an array `M:cv=0 12` and filled it with values. Calling the variable without index should give the mean value, but I am getting 0 instead.

```
;loop through cell voltages to create discharge tribble (DCCx)
for cntr 1 numcells 1
if (cv[cntr]-.05)>cv {
dcc|(1<<(cntr-1))
}
if (cv[cntr]<=cv {
dcc&(4095-(1<<(cntr)-1))
}
next

if upsecs%30==0 {
print DCC = %dcc%, MCI = %mci%, Min = %min%, Max = %max%, numcells = %numcells%, cv = %cv%
}
```
```
21:56:26.407 DCC = 0.00, MCI = 4095.00, Min = 3.69, Max = 3.80, numcells = 12.00, cv = 0.00
```
I have manually altered cell voltages to create a condition where cv[x] is greater than .05 from the mean, but the DCC byte is not altered. If I call cv from the console, it gives 0 instead of 3.72ish.

Note: If I change to median `m:cv=0 12` I get the correct value back.

**Reply by gemu2015** [reply]
you may not mix up a filter with an array.
a filter is an entity to either flatten a noisy value or remove outliers.
inside a filter there is always an array that is used for incoming values.
and therefore you may use a defined filter as an array as you already did.

but a real filter may not be used as an array. in your case if you want to smooth 12 cells you will 12 separate filters.

this demonstrates filter usage

\>D

M:movav=0 8
m:median=0 5

\>B
movav=1
movav=2
movav=3
movav=4
movav=5
movav=6
movav=7
movav=8

print moving average = %movav%

median=1
median=2
median=3
median=4
median=500

print median = %median%

**Reply by pkkrusty** [reply]
How to differentiate between filter and array in code? Looks the same to me.

I don't need to smooth values, I just want to use the median/mean capability of filters while still using the array to store individual values.

Which seems to work with `m:,` if I do
```
>D
m:cvr=0 18
m:cfr=0 7
m:cv=0 12
```
and load up the `cv` filter/array with values:
```
;Bit shift the 18x 8-bit registers to form the 12x 12-bit integers for each cell voltage reading then (ADC value * 1.5mV) to get voltage
for cntr 1 (numcells-1) 2
cv[cntr]=((cvr[cntr*1.5+0.5]%16)<<8)|(cvr[(cntr*1.5-0.5)]&255)*.0015
next
for cntr 2 numcells 2
cv[cntr]=(cvr[(cntr*1.5)]*16)|((cvr[(cntr*1.5-1)]>>4)&15)*.0015
next
```
And then try to use the median/Mean to compare indexed values against:
```
;loop through cell voltages to create discharge tribble (DCCx)
for cntr 1 numcells 1
temp2=cv[cntr]-0.05
if temp2>cv {
vstr+="d"
dcc=dcc|(1<<(cntr-1))
}
else { 
if cv[cntr]<=cv {
dcc=dcc&(4095-(1<<cntr-1))
}
vstr+="r"
}
next
```
I get this: 
```
14:33:34.439 DCC=3584.00, MCI=4095.00, cv=3.76, avg=3.75, temp1=1.00, temp2=3.81, vstr=rrrrrrrddd
```
But when I define `M:cv=0 12` I get:
```
15:06:04.409 DCC=4095.00, MCI=4095.00, cv=0.00, avg=3.75, temp1=99.00, temp2=3.81, vstr=dddddddddddd
```
I can manually average the 12 cells via
```
;average cell voltage values
avg=0
for cntr 1 numcells 1
avg+=cv[cntr]
next
avg/=numcells
```
but wanted to try saving some space since I thought the filter/array was constantly performing a running average.

**Reply by gemu2015** [reply]
the moving average filter keeps an index and a running sum, these are updated when you store a value in the unindexed variable.
so if you store indexed values in this variable they are not updated and will be wrong.

since you simply need an average over all cells i introduced an array average convenience function.
simply call array[-2] to get the average.

will do a pr later today

**Reply by pkkrusty** [reply]
Thanks for the PR. Reading more about SPI, and wondering about the clock speed for scripter. Most other tasmota code looks like 1Mhz, and some 4.5Mhz, but trying to calculate delay for my commands, which would be appreciable if speed is 115200 or less. Data sheet indicates at 1Mhz, will take less than 200µs to read the 18 cell voltage registers, so not an issue at that speed.

Also looks like tasmota always uses SPI mode 0?

Also wondering about the error correction byte that is sent after reading registers. If I pass a 19-cell array `cv` (1 command byte plus 18 voltage bytes) to get cell voltages, the chip normally also responds with a 20th byte for check-sum. Will that overflow into `cv[1]` or just not be read by the controller?

**Reply by gemu2015** [reply]
if you use software spi you may use any pins, also on an ESP8266, the speed being depended on cpu clock. you will get about 100kHz with software spi. if you need faster transfers you must use hardware spi. her you can specify the desired spi clock frequency.

most devices work with SPI Mode 0, so this is default.

SPI cannot receive without sending. to receive the last byte you must use a 20 bytes array.

**Reply by pkkrusty** [reply]
According to datasheet, 6802 uses SPI mode 3 (phase and polarity both 1). I found line 4350 in scripter.ino `glob_script_mem.spi.settings = SPISettings(fvar, MSBFIRST, SPI_MODE0);` that looks like it changes settings with hardware SPI defined in config, can I just change it to SPI_MODE3?

CS must be pulled low on writes and reads for the duration of the transaction, not sure what default is for scripting. But looks like that happens in the code.

**Reply by gemu2015** [reply]
yes you can. however most devices designed for mode3 also work with mode 0 and vice versa

**Reply by pkkrusty** [reply]
@gemu2015 New problem (after waiting for a logic analyzer to come from China). When I define the hardware SPI (in order to use the faster clock speed) in config page,
<img width="381" alt="Screen Shot 2022-08-06 at 12 52 12" src="https://user-images.githubusercontent.com/79770016/183249665-8b9ec504-e150-4b6a-89b6-8d3706c9fd53.png">
 and define hardware SPI in script with `res=spi(0 -1 1)` the driver responds with `warning: pins already used` and my logic analyzer shows that nothing is happening when the script runs, 
<img width="322" alt="Screen Shot 2022-08-06 at 12 53 23" src="https://user-images.githubusercontent.com/79770016/183249710-efbeb8eb-3215-4d6f-b695-461f4dcbda84.png">
besides a blip on the CS line every second.
If I define a software SPI (on the same pins), I get a little further, in that the chip select and clock line activates, and I get two bytes sent once every second (within 1ms)
<img width="1309" alt="Screen Shot 2022-08-06 at 12 57 12" src="https://user-images.githubusercontent.com/79770016/183249863-f19d4403-d646-4894-a02a-32abdd373924.png">

But in the >S section I do 
```
;spi command start cell measurements (takes 21ms) with discharge permitted (otherwise send 16)
res=spi(2 1 cmnd 1 1)
;delay(21)
res=spi(2 1 cvr 19 1)
;read cell voltage registers
```
where cmnd is 96 and first byte of cvr is 4. Nothing seems to be sent.

Any ideas where I'm going wrong? 

Of note, seems like the clock is handling 1mhz ok (I built with 160mhz mode).

**Reply by pkkrusty** [reply]
I was using 11.1.0.3 for these tests, but when I updated to 12.1.0.2, I get an exception for any SPI command other than spi(0... and spi(1...

```
{"Exception":28,"Reason":"Exception","EPC":["40212800","00000000","00000000"],"EXCVADDR":"00000000","DEPC":"00000000","CallChain":["402127f1","40234e41","40232a18","402330f3","4022f3d9","4022f3d9","401010d7","4010135f","402121a9","40247c9d","40247c9d","4024c120","4023317d","4022b898","4023c819","400042db","40004b31","401069c8","40244c38","40014000","40034002","4024c120","4022cfb8","40235270","401009b6","4023555d","4024ec05","401009b6","40219b32","4023ce17","4023ce4c"]},"BootCount":3}
```

Maybe due to changes introduced in https://github.com/arendst/Tasmota/commit/cc771f31e9f1d814d11ad9adf3d91f78ea672834 ?

**Reply by gemu2015** [reply]
i testet with current dev version (esp32 version, which i had under test anyhow)
and it works like expected. i expect ESP8266 to work also
for hardware SPI you must define MOSI, MISO,CLK in Tasmota WEB GUI but no CS
CS is defined via SPI(1 1 pin)

according to the crash please give a minimal script with which i can try to reproduce.

**Reply by pkkrusty** [reply]
Here is key part of script that crashes for me.  If I comment out spi(2... it doesn't crash. Maybe has to do with defining pin 5 as CS. I will try your suggestion tomorrow.
```
>D
m:cvr=0 19
m:cfr=0 7
m:wcfr=0 7
m:cv=0 12
res=0
cmnd=96

vstr=""

>B
;put some values in cvr so script can test bit shifting the registers to get cell voltages
cvr[0]=0
cvr=4
cvr=123
cvr=137
cvr=152
cvr=149
cvr=57
cvr=154
cvr=176
cvr=217
cvr=155
cvr=203
cvr=137
cvr=157
cvr=229
cvr=57
cvr=159
cvr=0
cvr=218
cvr=160
;Comparator duty cycle 3 bits, default 5, with other bits, WDT, GPIO1, GPIO2, LVLPL, CELL10 all 0, so config register [0] is just 5
wcfr=1
wcfr=5
res=spi(0 -1 1)
res=spi(1 1 5)

>S
;cv[1]=((cvr[2]%16)<<8)|(cvr[1]&255)*.0015
;cv[2]=(cvr[3]*16)|((cvr[2]>>4)&15)*.0015

;spi command start cell measurements (takes 21ms) with discharge permitted (otherwise send 16)
res=spi(2 1 cmnd 1 1)
delay(21)
res=spi(2 1 cvr 19 1)
;read cell voltage registers
```

**Reply by gemu2015** [reply]
your script crashes because you try to send a variable, but you must send an array
i tried on an esp8266 and it works as expected if you leave this line
res=spi(2 1 cmnd 1 1)

**Reply by pkkrusty** [reply]
Yep, that was it. Works as expected when I send an array instead. Thank you. Of note, I can only make it work if I define all four SPI pins in GUI config page (CS included).

**Reply by gemu2015** [reply]
ah yes the esp8266 enforces cs or dc, so you have to define cs and ignore the double define warning
esp32 does not.

**Reply by pkkrusty** [reply]
@gemu2015 Now working with a BMS (ISL78600) that needs CS pulled high between each byte, is that something you can add with a #define? 
<img width="552" alt="Screen Shot 2023-03-05 at 20 57 23" src="https://user-images.githubusercontent.com/79770016/222985551-3ba859eb-be95-4f5f-980d-b1887ec7dc2e.png">

**Reply by gemu2015** [reply]
i added this option by using transfer size S = 4 (new version see my fork)

spi(2 N ARRAY LEN S) sends and receives an ARRAY with size values with S (1..3) (8,16,24 bits) if N==-1 cs is
ignored

**Reply by pkkrusty** [reply]
Wonderful, much thanks!

**Reply by pkkrusty** [reply]
Is it possible to define multiple CS with ESP8266? can I just do four definition commands:
```
res=spi(1 1 1)
res=spi(1 2 0)
res=spi(1 3 16)
res=spi(1 4 2)
```
and then pass arrays like normal:
```
cs=upsecs%4+1
res=spi(2 cs rdcmd 4 4)
```
I have to define CS in the GPIO definitions for ESP8266 before it will work, but maybe I can just define all four of those as CS?
<img width="397" alt="Screen Shot 2023-03-08 at 16 08 16" src="https://user-images.githubusercontent.com/79770016/223766429-19876a6e-4c5c-4762-ae3a-0f96156de89e.png">

-------

Another mode of the ISL78600 uses a DATAREADY line to signal the ESP8266 that data can be clocked out (one byte at a time). Is there any provision for that in your SPI implementation? I'm guessing no, since it's not in the docs.

**Reply by gemu2015** [reply]
up to 4 cs lines are supported thus 4 different chips may be addressed.
Tasmota requires to define the cs line to init hardware SPI, but you need only 1 CS defined.

DATAREADY mode is not supported

**Comment by gemu2015** [comment]
@laurentppol 

to have a template for serial asci queries i adapted your meter. please try and modify as required. i will then put this template to the docs for other users.

\>D 128
;read JUNCTEC KG-F meter
; should be compiled with these additional defines
;#define USE_SCRIPT_SERIAL
;#define MAXSVARS 4
;#define SCRIPT_MAXSSIZE 128

res=0

volt=0
curr=0
rcap=0
ccap=0
whr=0
rtim=0
temp=0
dpwr=0
stat=0
dir=0
blife=0
bres=0

rstr=""

\>B
; open serial port, replace pin numbers with your actual wiring recpin, trxpin and baudrate
; should be hardware serial for 115200
res=so(3 1 115200 8N1 1024)

\>S
; check if serial port received bytes
res=sa()
if res>0 {
; read until linefeed
rstr=sr(10)

;:r50=2,215,2056,200, 5408,4592,9437,14353 , 134,4112,0,0,162,3068 2,
; scale each value as needed
volt=st(rstr , 3)
volt/=100
curr=st(rstr , 4)
curr/=100
rcap=st(rstr , 5)
rcap/=1000
ccap=st(rstr , 6)
ccap/=1000
whr=st(rstr , 7)
whr/=100000
rtim=st(rstr , 8)
temp=st(rstr , 9)
dpwr=st(rstr , 10)
dpwr/=100
stat=st(rstr , 11)
dir=st(rstr , 12)
blife=st(rstr , 13)
bres=st(rstr , 14)
bres/=10
}

; transmit cmd
;:R50=2,2, 1,
res=sw(":R50=2,2, 1,\r\n")

; MQTT export
\>J
,"BMS":{
"volt":%2volt%,
"curr":%2curr%,
"rcap":%2rcap%,
"ccap":%2ccap%,
"whr":%2whr%,
"rtim":%2rtim%,
"temp":%2temp%,
"dpwr":%2dpwr%,
"stat":%0stat%,
"dir":%0dir%,
"blife":%2blife%,
"bres":%2bres%,


; web display
\>W
Voltage{m}%2volt% V
Current{m}%2curr% A
R-Capacity{m}%4rcap% Ah
C-Capacity{m}%4ccap% Ah
Watt-hour{m}%5whr% Wh
Running time{m}%0rtim% s
Temperature{m}%0temp% C
Power{m}%2dpwr% W
Status{m}%0stat%
Direction{m}%0dir%
Battery life{m}%0blife% mins
Battery resistance{m}%2bres% mOhm

**Reply by laurentppol** [reply]
THX a lot, will try.

**Reply by oli82** [reply]
@gemu2015 did you have your original script to read the JUNCTEC KG-F meter and can you send it to me?

Thx

**Comment by laurentppol** [comment]
Ok, compile with #define suggested, but now, HOW do I upload script to   NodeMCU?

My user_config_override.h:
`
#define MY_LANGUAGE            pl_PL           // Polish in Poland

#define USE_BH1750                             // [I2cDriver11] Enable BH1750 sensor (I2C address 0x23 or 0x5C) (+0k5 code)

#define USE_INA226                             // [I2cDriver35] Enable INA226 (I2C address 0x40, 0x41 0x44 or 0x45) Low voltage and current sensor (+2k3 code)

#define USE_ADS1115                            // [I2cDriver13] Enable ADS1115 16 bit A/D converter (I2C address 0x48, 0x49, 0x4A or 0x4B) based on Adafruit ADS1x15 library (no library needed) (+0k7 code)

//#define USE_WE517                                // Add support for Orno WE517-Modbus energy monitor (+1k code)

#define USE_RC_SWITCH                            // Add support for RF transceiver using library RcSwitch (+2k7 code, 460 iram)

#undef USE_RULES
#define USE_SCRIPT
#define USE_SCRIPT_SERIAL
#define MAXSVARS 4
#define SCRIPT_MAXSSIZE 128
`

**Reply by jkwim** [reply]
You need to flash the ESP with this firmware and then within Tasmota find Consoles->Script Editor. Type the script there.
